app/services/authentication/providers.rb

```diff
@@ -1,11 +1,69 @@
 module Authentication
   module Providers
-    # TODO: [thepracticaldev/oss] raise exception if provider is available but not enabled for this app
-    # TODO: [thepracticaldev/oss] add available providers, enabled providers
+    # Retrieves a provider that is both available and enabled
     def self.get!(provider_name)
-      "Authentication::Providers::#{provider_name.titleize}".constantize
-    rescue NameError => e
-      raise ::Authentication::Errors::ProviderNotFound, e
+      name = provider_name.to_s.titleize
+
+      "Authentication::Providers::#{name}".safe_constantize.tap do |klass|
```

Comment:
`safe_constantize` potentially raises and handles the error internally, which is relatively expensive for something that can be done without an exception:

```ruby
name = provider_name.to_s.titleize
unless Authentication::Providers.const_defined?(name)
  ...
end

unless enabled?(provider_name)
  ...
end

Authentication::Providers.const_get(name)
```

This way we can also eliminate the use of `tap`, which I'm not a big fan of in situations like this.

Reply:
Perfect, i'll change this too. Thanks a lot!